src/toil/wdl/wdl_types.py

```diff
@@ -1,41 +1,211 @@
+from abc import ABC
 from typing import Any
 
 
-class WDLType(str):
+class WDLRuntimeError(RuntimeError):
+    pass
+
+
+class WDLType:
     """
     Represents a primitive or compound WDL type:
 
     https://github.com/openwdl/wdl/blob/main/versions/development/SPEC.md#types
     """
-    # TODO: make this into its own class instead of extending string.
-    #  This requires all WDL types to be a WDLType.
+
+    def __init__(self, optional: bool = False):
+        self.optional = optional
+
+    @property
+    def name(self) -> str:
+        """
+        Type name as string. Used in display messages / 'mappings.out' if dev
+        mode is enabled.
+        """
+        raise NotImplementedError
+
+    @property
+    def default_value(self):
+        """
+        Default value if optional.
+        """
+        return None
+
+    def create(self, value: Any) -> Any:
+        """
+        Calls at runtime. Returns an instance of the current type. An error may
+        be raised if the value is not in the correct format.
+
+        :param value: a Python object
+        """
+        if value is None:
+            # check if input is in fact an optional.
+            if self.optional:
+                return self.default_value
+            else:
+                raise WDLRuntimeError(f"Required input for '{self.name}' type not specified.")
+
+        return self._create(value)
+
+    def _create(self, value: Any) -> Any:
+        raise NotImplementedError
+
+    def __eq__(self, other):
+        # TODO: remove after refactor
+        return self.name.__eq__(other)
```

Comment:
Is this line still needed, or is the refactor done enough to drop it?

Reply:
I decided to keep this since it makes sense to compare two types. All the types are fully refactored away from str, however.